### ISSUE TITLE
Slow CPU compile of unrolled autoregressive sampling loop

### ISSUE THREAD

**Issue by j-towns** [google/jax#1223]
__TLDR:__ The HLO linked at the bottom is very slow to compile.
***
I'm writing a tool, called [fastar](https://github.com/j-towns/fastar), to do fast sampling from autoregressive models using JAX. This style of sampling requires running a loop where the shapes of arrays change between loop iterations, thus it's not possible to use the lax loop primitives to run the loop. This sampling operation is CPU-friendly because it's basically a long series of small array operations. These computations also involve a lot of very sparse (often single element) scatter update ops.

I've been experimenting with jitting sections of the unrolled loop, with the eventual aim of hopefully being able to jit the entire thing. Compile times are slow on the (relatively small) example loop that I'm working with. This procedure samples an 8 x 8 image. The loop has one iteration per pixel-channel, so 8 x 8 x 3 = 192 iterations.

Compilation of the first 15 iterations takes about 6 minutes. I tried compiling a full 192 iteration loop and it hadn't compiled after about an hour on my machine. I've attached hlo for the 15 iteration computation and the full 192 iteration one.

Note that the image size and the model from which I'm sampling are both relatively small. If we think it's unrealistic to expect to be able to compile unrolled loops like this in a reasonable time then that would be good to know, and would not be a deal breaker, because I can work around by compiling sub-sections of the loop.

[15 iteration HLO (2.1 MB)](https://github.com/google/jax/files/3526231/module_1051.before_optimizations.txt)

[192 iteration HLO (60 MB)](https://drive.google.com/file/d/1vcvEWxnx-LKxgTH8L0ehNsBCD6WUotHY/view?usp=sharing)

FYI I've been using jaxlib version 0.1.23.

**Comment by j-towns** [comment]
@hawkinsp just wondering if you might have time to take a look at this. Would be great to get a rough idea of why this is so slow (simply the large number of ops, or that certain ops we’re using are particularly slow to compile, or some obvious performance issue in xla).

**Comment by Jeevesh8** [comment]
@j-towns What happened? I am dealing with loops of this kind too. And it takes me similar amounts of time too. How did you optimize(if you were able to)?

Your gratefully

**Comment by j-towns** [comment]
Basically, CPU compilation is generally significantly slower than GPU and TPU. I'm gonna close this for now because it's not really a relevant problem for me any more. @Jeevesh8 you might be able to speed up compilation by using JAX's functional control flow functions (specifically the loop ones): https://jax.readthedocs.io/en/latest/jax.lax.html#control-flow-operators.

**Comment by Jeevesh8** [comment]
Indeed I was able to speed up compilation by using Jax primitives. As it turns out that for loops will always be unrolled fully, and then optimized, whereas we can specify the amount of unroll in ``jax.lax.scan``. Thank you @j-towns .

But I observed something interesting. So initially, the function that I was ``jit/pmap``-ing was a function that had ``jax.value_and_grad`` within it. That is, I was trying to ``jit`` the operation of differentiating through the loop. 

After, I converted the loop to ``jax.lax.scan`` , the compile time was still over an hour. But, when I cahnged my code to `'jit`` first and then  take the ``jax.value_and_grad``, it compiled quite fast(7-8 minutes). 

So, my question now is: does XLA unroll the scan loop when it is wrapped in ``jax.value_and_grad`` and not otherwise? Or is it some other reason?

I checked the Jaxprs of  ``jit`` of ``grad`` of scan loop and they contain ``scan`` only. I wonder why this huge time difference then?
